File: 1146-greatest-common-divisor-of-strings/greatest-common-divisor-of-strings.js
/**
 * @param {string} str1
 * @param {string} str2
 * @return {string}
 */
var gcdOfStrings = function(str1, str2) {
    if (str1 + str2 !== str2 + str1) {
        return ""
    }

    let remainder;
    let lengthOfStr1 = str1.length
    let lengthOfStr2 = str2.length

    remainder = lengthOfStr1 % lengthOfStr2

    while (remainder != 0) {
        const tmpRemainder = remainder
        lengthOfStr1 = lengthOfStr2
        remainder = lengthOfStr1 % tmpRemainder
        lengthOfStr2 = tmpRemainder
    }

    return str1.slice(0,lengthOfStr2)
};

File: 1146-greatest-common-divisor-of-strings/greatest-common-divisor-of-strings.rb
# @param {String} str1
# @param {String} str2
# @return {String}
def gcd_of_strings(str1, str2)
    return "" unless str1 + str2 == str2 + str1

    a = str1.length
    b = str2.length

    gcd = a.gcd(b)
    str1[0...gcd]
end